### ISSUE TITLE
X: Cannot type unicode characters without corresponding keysym

### ISSUE THREAD

**Issue by nzahasan** [moses-palmer/pynput#94]
I was trying to type Bangla unicode character, but it fails with this error

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/pynput/keyboard/_base.py", line 450, in type
    self.release(character)
  File "/usr/local/lib/python3.6/dist-packages/pynput/keyboard/_base.py", line 404, in release
    self._handle(resolved, False)
  File "/usr/local/lib/python3.6/dist-packages/pynput/keyboard/_xorg.py", line 196, in _handle
    raise self.InvalidKeyException(key)
pynput.keyboard._base.InvalidKeyException: 'খ'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "pyinp.py", line 6, in <module>
    virtualKB.type("I am bad খ")
  File "/usr/local/lib/python3.6/dist-packages/pynput/keyboard/_base.py", line 453, in type
    raise self.InvalidCharacterException(i, character)
pynput.keyboard._base.InvalidCharacterException: (9, 'খ')
```

**Comment by moses-palmer** [comment]
Thank you for your report.

Could you please include the full string you're trying to type?

**Comment by moses-palmer** [comment]
It appears that the letter you are trying to type does not have a corresponding X keysymdef as can be seen when searching for it in [xorg_keysyms.py](https://github.com/moses-palmer/pynput/blob/master/lib/pynput/_util/xorg_keysyms.py.

What output do you get when running `xev` and pressing the key you use to type _BENGALI LETTER KHA_?

**Comment by nzahasan** [comment]
I do not have a BANGLA KHA in my keyboard. I my intention is to press k and type BANGLA KHA .

**Comment by Evidlo** [comment]
So what's the right solution for typing characters for which there is no keysym?  I'm having the same problem with `⏎`.

**Comment by nzahasan** [comment]
apparently you have to add the character in [xorg_keysyms.py](https://github.com/moses-palmer/pynput/blob/master/lib/pynput/_util/xorg_keysyms.py) .

**Comment by moses-palmer** [comment]
That is not entirely correct, as that file is generated by parsing the actual _Xlib_ header files. These files, unfortunately, do not contain definitions for _RETURN SYMBOL_ and _BENGALI LETTER KHA_.

Does anybody know whether `xdotool` is able to type these characters?

**Comment by Evidlo** [comment]
xdotool can type the return symbol, at least.

**Comment by moses-palmer** [comment]
@nzahasan @Evidlo If you run `xev` at the same time as typing these characters with `xdotool`, what events are displayed in the terminal?

**Comment by Evidlo** [comment]
```
MappingNotify event, serial 53, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 53, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 53, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 53, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 53, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 53, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 53, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 53, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 53, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 57, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 248

MappingNotify event, serial 63, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 63, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 63, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 63, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 63, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 63, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 63, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 63, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 63, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 1

MappingNotify event, serial 72, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 248

```

**Comment by gdiShun** [comment]
To resurrect this, I was able to remap Caps Lock to a unicode emoji with xmodmap like this:
`xmodmap -e "keycode  66 = U0001F61A"`

xev output:
```
KeyPress event, serial 37, synthetic NO, window 0x6800001,
    root 0x6bf, subw 0x0, time 15707692, (11,-24), root:(1842,985),
    state 0x12, keycode 66 (keysym 0x101f61a, U0001F61A), same_screen YES,
    XLookupString gives 4 bytes: (f0 9f 98 9a) "😚"
    XmbLookupString gives 4 bytes: (f0 9f 98 9a) "😚"
    XFilterEvent returns: False

KeyRelease event, serial 37, synthetic NO, window 0x6800001,
    root 0x6bf, subw 0x0, time 15707774, (11,-24), root:(1842,985),
    state 0x12, keycode 66 (keysym 0x101f61a, U0001F61A), same_screen YES,
    XLookupString gives 4 bytes: (f0 9f 98 9a) "😚"
    XFilterEvent returns: False
```

**Comment by moses-palmer** [comment]
@gdiShun, thank you for your thorough analysis!

I downloaded the source for `xmodmap` to check what it does. `keycode kc = ks` will eventually cause to to pass `ks` to `XStringToKeysym`. This is a utility function defined in C and is not available to the Python library `Xlib`, but for strings starting with `'U'` it will perform the following:

```c
    if (*s == 'U') {
        val = 0;
        for (p = &s[1]; *p; p++) {
            c = *p;
            if ('0' <= c && c <= '9') val = (val<<4)+c-'0';
            else if ('a' <= c && c <= 'f') val = (val<<4)+c-'a'+10;
            else if ('A' <= c && c <= 'F') val = (val<<4)+c-'A'+10;
            else return NoSymbol;
            if (val > 0x10ffff)
                return NoSymbol;
        }
        if (val < 0x20 || (val > 0x7e && val < 0xa0))
            return NoSymbol;
        if (val < 0x100)
            return val;
        return val | 0x01000000;
    }
```

This is where `0x100002c` comes from. As you note, simply passing this as a keysym does not help. It will then call `XChangeKeyboardMapping` to remap the keyboard. Perhaps this is what makes it work? In your experiments above, did you attempt to have *pynput* type `'😚'` after having remapped it? Did you try with the following patch applied?

```diff
diff --git a/lib/pynput/_util/xorg_keysyms.py b/lib/pynput/_util/xorg_keysyms.py
index db501ff..ee61ead 100644
--- a/lib/pynput/_util/xorg_keysyms.py
+++ b/lib/pynput/_util/xorg_keysyms.py
@@ -18,6 +18,7 @@
 # pylint: disable=C0111,C0302
 
 SYMBOLS = {
+    'Kissing_Face_with_Closed_Eye': (0x0101F61A, u'\u1F61A'),
     '0': (0x0030, u'\u0030'),
     '1': (0x0031, u'\u0031'),
     '2': (0x0032, u'\u0032'),
```

**Comment by gdiShun** [comment]
> In your experiments above, did you attempt to have pynput type '😚' after having remapped it?

I tried a lot of things, so I can't say with 100% certainty that I did, but I'm pretty sure I did. The result was likely just no output. What I do clearly remember doing is using `from_vk(0x0001f61a)`. It would return the proper character (`'😚'`) when print()'d. pynput's type would only return `''`. And pynput's tap would give a similar error as the last one in my previous post, until I added it to Xlib's keysymdef. Then it would once again do nothing.

> Did you try with the following patch applied?

The patch as-is causes the same `pynput.keyboard._base.InvalidCharacterException` error. The `u'\u1F61A'` needs to be replaced with `u'\U0001F61A'` for no errors. But still no output. Also tried adding `XK_Kissing_Face_with_Closed_Eye = 0x1F61A`(and other variations) to the Xlib keysymdef. And again, no errors, but no output either. Interestingly, even with all these keysymdef changes, using `tap(Key.caps_lock)` after mapping it will still cause that same `BadValue` error.

I'm thinking this might be more font-, or software-based than input-, or hardware-based. More of a hunch than anything. Really don't know what I'm talking about here. :p But I've been scouring for a library to do this, and root-access or not, I can't find any that will actually output emojis or other non-standard unicode characters.

As a bit of a side-note, that may be helpful. I found a solution to inputting these characters, not with pynput, but evdev. A hacky solution using the CTRL+SHIFT+U shortcut. I posted it [here](https://stackoverflow.com/questions/62695914/how-to-type-utf-8-emojis-with-the-pynput-keyboard-module).What's interesting to me, is that this works great with evdev, but inputting the same CTRL+SHIFT+U shortcut on pynput does not work either. I'm guessing it has to do with how evdev injects inputs directly?

**Comment by moses-palmer** [comment]
I have pushed a proposed solution to [`fixup-xorg-chars-without-keysyms`](https://github.com/moses-palmer/pynput/tree/fixup-xorg-chars-without-keysyms). It actually appears to work on my system, but I would like confirmation from any of @Evidlo, @gdiShun or @nzahasan before merging into *master*.

**Comment by gdiShun** [comment]
```
Controller().type("'😚'\N{kissing face with closed eyes}\U0001F61A")
Controller().tap('😚')
```
'😚'😚😚😚'😚'😚😚😚'😚'😚😚😚'😚'😚😚😚

Working!

**Comment by moses-palmer** [comment]
This has been released as *pynput 1.7*.
